Pcap++/header/KniDevice.h

```diff
@@ -0,0 +1,709 @@
+		 */
+		struct KniIoctlCallbacks
+		{
+			/**
+			 * Pointer to function of changing MTU.
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*change_mtu)(uint16_t port_id, unsigned int new_mtu);
+			/**
+			 * Pointer to function of configuring network interface.
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*config_network_if)(uint16_t port_id, uint8_t if_up);
+			/**
+			 * Pointer to function of configuring mac address.
+			 * If callback is not provided and port_id of KNI device is not UINT16_MAX
+			 * then DPDK will use predefined callback witch sets MAC address of
+			 * DPDK port port_id via rte_eth_dev_default_mac_addr_set.
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*config_mac_address)(uint16_t port_id, uint8_t mac_addr[]);
+			/**
+			 * Pointer to function of configuring promiscuous mode.
+			 * If callback is not provided and port_id of KNI device is not UINT16_MAX
+			 * then DPDK will use predefined callback witch sets promiscuous mode of
+			 * DPDK port port_id via rte_eth_promiscuous_enable/rte_eth_promiscuous_disable.
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*config_promiscusity)(uint16_t port_id, uint8_t to_on);
+		};
+
+		/**
+		 * @brief Old callbacks for KNI device events.
+		 * This structure MUST be used ONLY when KniDevice#callbackVersion returns
+		 * KniCallbackVersion#CALLBACKS_OLD.
+		 * Or if You are sure that DPDK version used is lower than 17.11.
+		 * If some callback is not provided (NULL) the request will always succeeds.
+		 */
+		struct KniOldIoctlCallbacks
+		{
+			/**
+			 * Pointer to function of changing MTU.
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*change_mtu)(uint8_t port_id, unsigned int new_mtu);
+			/**
+			 * Pointer to function of configuring network interface.
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*config_network_if)(uint8_t port_id, uint8_t if_up);
+		};
+
+		/**
+		 * @brief KNI device initialization data.
+		 * Used to create new KNI device.
+		 * Usage of callbacks member or oldCallbacks member is defined by
+		 * result of KniDevice#callbackVersion
+		 */
+		struct KniDeviceConfiguration
+		{
+			enum
+			{
+				// Must correspond to minimum of
+				// RTE_KNI_NAMESIZE [DPDK] = 32 and
+				// IFNAMSIZ [LINUX] = 16 (on most systems)
+				KNI_NAME_SIZE = 16
+			};
+			/**
+			 * Name used to display device in system.
+			 * Must not interfere with already existing network interfaces.
+			 */
+			char name[KNI_NAME_SIZE];
+			union
+			{
+				KniIoctlCallbacks* callbacks;
+				KniOldIoctlCallbacks* oldCallbacks;
+			};
+			/**
+			 * Pointer to MAC (ETHERNET) address of new KNI device.
+			 * If omitted (NULL) some valid address will be automatically generated.
+			 * If provided will be cached by new KNI device info structure.
+			 */
+			MacAddress* mac;
```

Comment:
This is not a place for dynamicaly allocated object (see tests where the address of local variable is passed to this structure). It is a simple replacement for std::optional -> if nullptr then use some default if not nullptr use waht is provided
The valid MAC for KNI will be automatically generated by DPDK in case if it is not provided.
This pointer is used only inside KniDevice constructor for DPDK the copy of MAC is provided

Reply:
I agree that this ptr is only used inside of `KniDevice` d'tor but still it requires the user of this class to remember that he put this pointer to an object he owns. Since this specific code isn't running in real-time I think it'd be better if this instance is owned by `KniDeviceConfiguration` and you can initialize it with `MacAddress::Zero` which is like having a NULL

Reply:
Ok maybe I am overwhelmed with the desire to optimize everyting so this most cerntaly will not be a use case where the thousands of KNI devices will be created rapidly so small tweak from C to the more C++ like approach is reasonable here.
This MUST be stated in documentation and tested:
> You can initialize this member with MacAddress::Zero which can be used as an equivalent to NULL

I will close similar discussions about this structure and let us continue only in this thread.

Reply:
👍

Reply:
\+